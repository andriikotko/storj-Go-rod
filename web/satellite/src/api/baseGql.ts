// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import gql from 'graphql-tag';

import { apollo } from '@/utils/apollo';

/**
 * BaseGql is a graphql utility which allows to perform queries and mutations.
 */
export class BaseGql {
    /**
     * performs qraphql query.
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async query(query: string, variables: any = null): Promise<any> {
        const response: any = await apollo.query(
            {
                query: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            },
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    /**
     * performs qraphql mutation.
     *
     * @param query - qraphql query
     * @param variables - variables to bind in query. null by default.
     * @throws Error
     */
    protected async mutate(query: string, variables: any = null): Promise<any> {
        const response: any = await apollo.mutate(
            {
                mutation: gql(query),
                variables,
                fetchPolicy: 'no-cache',
                errorPolicy: 'all',
            },
        );

        if (response.errors) {
            throw new Error(this.combineErrors(response.errors));
        }

        return response;
    }

    private combineErrors(gqlError: any): string {
        return gqlError.map(err => err).join('\n');
    }
}
