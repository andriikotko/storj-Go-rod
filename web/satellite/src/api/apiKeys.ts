// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ApiKey, ApiKeyCursor, ApiKeysApi, ApiKeysPage } from '@/types/apiKeys';

/**
 * ApiKeysApiGql is a graphql implementation of ApiKeys API.
 * Exposes all apiKey-related functionality
 */
export class ApiKeysApiGql extends BaseGql implements ApiKeysApi {
    /**
     * Fetch apiKeys.
     *
     * @returns ApiKey
     * @throws Error
     */
    public async get(projectId: string, cursor: ApiKeyCursor): Promise<ApiKeysPage> {
        const query =
            `query($projectId: String!, $limit: Int!, $search: String!, $page: Int!, $order: Int!, $orderDirection: Int!) {
                project (
                    id: $projectId,
                ) {
                    apiKeys (
                        cursor: {
                            limit: $limit,
                            search: $search,
                            page: $page,
                            order: $order,
                            orderDirection: $orderDirection
                        }
                    ) {
                        apiKeys {
                            id,
                            name,
                            createdAt
                        }
                        search,
                        limit,
                        order,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId: projectId,
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
            order: cursor.order,
            orderDirection: cursor.orderDirection,
        };

        const response = await this.query(query, variables);

        return this.getApiKeysPage(response.data.project.apiKeys);
    }

    /**
     * Used to create apiKey.
     *
     * @param projectId - stores current project id
     * @param name - name of apiKey that will be created
     * @returns ApiKey
     * @throws Error
     */
    public async create(projectId: string, name: string): Promise<ApiKey> {
        const query =
            `mutation($projectId: String!, $name: String!) {
                createAPIKey(
                    projectID: $projectId,
                    name: $name
                ) {
                    key,
                    keyInfo {
                        id,
                        name,
                        createdAt
                    }
                }
            }`;

        const variables = {
            projectId,
            name,
        };

        const response = await this.mutate(query, variables);
        const key: any = response.data.createAPIKey.keyInfo;
        const secret: string = response.data.createAPIKey.key;

        return new ApiKey(key.id, key.name, key.createdAt, secret);
    }

    /**
     * Used to delete apiKey.
     *
     * @param ids - ids of apiKeys that will be deleted
     * @throws Error
     */
    public async delete(ids: string[]): Promise<void> {
        const query =
            `mutation($id: [String!]!) {
                deleteAPIKeys(id: $id) {
                    id
                }
            }`;

        const variables = {
            id: ids,
        };

        const response = await this.mutate(query, variables);

        return response.data.deleteAPIKeys;
    }

    /**
     * Method for mapping api keys page from json to ApiKeysPage type.
     *
     * @param page anonymous object from json
     */
    private getApiKeysPage(page: any): ApiKeysPage {
        if (!page) {
            return new ApiKeysPage();
        }

        const apiKeysPage: ApiKeysPage = new ApiKeysPage();

        apiKeysPage.apiKeys = page.apiKeys.map(key => new ApiKey(
            key.id,
            key.name,
            new Date(key.createdAt),
            '',
        ));

        apiKeysPage.search = page.search;
        apiKeysPage.limit = page.limit;
        apiKeysPage.order = page.order;
        apiKeysPage.orderDirection = page.orderDirection;
        apiKeysPage.pageCount = page.pageCount;
        apiKeysPage.currentPage = page.currentPage;
        apiKeysPage.totalCount = page.totalCount;

        return apiKeysPage;
    }
}
