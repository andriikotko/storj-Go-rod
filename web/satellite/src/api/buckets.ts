// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { Bucket, BucketCursor, BucketPage, BucketsApi } from '@/types/buckets';

/**
 * BucketsApiGql is a graphql implementation of Buckets API.
 * Exposes all bucket-related functionality.
 */
export class BucketsApiGql extends BaseGql implements BucketsApi {
    /**
     * Fetch buckets.
     *
     * @returns BucketPage
     * @throws Error
     */
    public async get(projectId: string, before: Date, cursor: BucketCursor): Promise<BucketPage> {
        const query =
            `query($projectId: String!, $before: DateTime!, $limit: Int!, $search: String!, $page: Int!) {
                project(id: $projectId) {
                    bucketUsages(before: $before, cursor: {
                        limit: $limit, search: $search, page: $page
                    }) {
                        bucketUsages {
                            bucketName,
                            storage,
                            egress,
                            objectCount,
                            since,
                            before
                        },
                        search,
                        limit,
                        offset,
                        pageCount,
                        currentPage,
                        totalCount
                    }
                }
            }`;

        const variables = {
            projectId,
            before: before.toISOString(),
            limit: cursor.limit,
            search: cursor.search,
            page: cursor.page,
        };

        const response = await this.query(query, variables);

        return this.getBucketPage(response.data.project.bucketUsages);
    }

    /**
     * Method for mapping buckets page from json to BucketPage type.
     *
     * @param page anonymous object from json
     */
    private getBucketPage(page: any): BucketPage {
        if (!page) {
            return new BucketPage();
        }

        const buckets: Bucket[] = page.bucketUsages.map(key =>
            new Bucket(
                key.bucketName,
                key.storage,
                key.egress,
                key.objectCount,
                new Date(key.since),
                new Date(key.before)));

        return new BucketPage(buckets, page.search, page.limit, page.offset, page.pageCount, page.currentPage, page.totalCount);
    }
}
