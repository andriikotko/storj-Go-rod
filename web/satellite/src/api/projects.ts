// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { BaseGql } from '@/api/baseGql';
import { ErrorUnauthorized } from '@/api/errors/ErrorUnauthorized';
import { Project, ProjectFields, ProjectLimits, ProjectsApi } from '@/types/projects';
import { HttpClient } from '@/utils/httpClient';

export class ProjectsApiGql extends BaseGql implements ProjectsApi {
    private readonly http: HttpClient = new HttpClient();
    private readonly ROOT_PATH: string = '/api/v0/projects';

    /**
     * Creates project.
     *
     * @param projectFields - contains project information
     * @throws Error
     */
    public async create(projectFields: ProjectFields): Promise<Project> {
        const query =
            `mutation($name: String!, $description: String!) {
                createProject(
                    input: {
                        name: $name,
                        description: $description,
                    }
                ) {id}
            }`;

        const variables = {
            name: projectFields.name,
            description: projectFields.description,
        };

        const response = await this.mutate(query, variables);

        return new Project(response.data.createProject.id, variables.name, variables.description, '', projectFields.ownerId);
    }

    /**
     * Fetch projects.
     *
     * @returns Project[]
     * @throws Error
     */
    public async get(): Promise<Project[]> {
        const query = `query {
            myProjects{
                name
                id
                description
                createdAt
                ownerId
            }
        }`;

        const response = await this.query(query);

        return response.data.myProjects.map((project: Project) => {
            return new Project(
                project.id,
                project.name,
                project.description,
                project.createdAt,
                project.ownerId,
            );
        });
    }

    /**
     * Update project name and description.
     *
     * @param projectId - project ID
     * @param name - project name
     * @param description - project description
     * @returns Project[]
     * @throws Error
     */
    public async update(projectId: string, name: string, description: string): Promise<void> {
        const query =
            `mutation($projectId: String!, $name: String!, $description: String!) {
                updateProject(
                    id: $projectId,
                    name: $name,
                    description: $description
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
            name: name,
            description: description,
        };

        await this.mutate(query, variables);
    }

    /**
     * Delete project.
     *
     * @param projectId - project ID
     * @throws Error
     */
    public async delete(projectId: string): Promise<void> {
        const query =
            `mutation($projectId: String!) {
                deleteProject(
                    id: $projectId
                ) {name}
            }`;

        const variables = {
            projectId: projectId,
        };

        await this.mutate(query, variables);
    }

    /**
     * Get project limits.
     *
     * @param projectId- project ID
     * throws Error
     */
    public async getLimits(projectId): Promise<ProjectLimits> {
        const path = `${this.ROOT_PATH}/${projectId}/usage-limits`;
        const response = await this.http.get(path);

        if (response.ok) {
            const limits = await response.json();

            return new ProjectLimits(
                limits.bandwidthLimit,
                limits.bandwidthUsed,
                limits.storageLimit,
                limits.storageUsed,
            );
        }

        if (response.status === 401) {
            throw new ErrorUnauthorized();
        }

        throw new Error('can not get usage limits');
    }
}
