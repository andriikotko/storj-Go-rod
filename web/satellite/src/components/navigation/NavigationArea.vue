// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="navigation-area" v-if="!isNavigationHidden">
        <EditProjectDropdown/>
        <router-link
            :aria-label="navItem.name"
            class="navigation-area__item-container"
            v-for="navItem in navigation"
            :key="navItem.name"
            :to="navItem.path"
        >
            <div class="navigation-area__item-container__link">
                <component :is="navItem.icon"></component>
                <h1 class="navigation-area__item-container__link__title">{{navItem.name}}</h1>
            </div>
        </router-link>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import EditProjectDropdown from '@/components/navigation/EditProjectDropdown.vue';

import ApiKeysIcon from '@/../static/images/navigation/apiKeys.svg';
import DashboardIcon from '@/../static/images/navigation/dashboard.svg';
import TeamIcon from '@/../static/images/navigation/team.svg';

import { RouteConfig } from '@/router';
import { NavigationLink } from '@/types/navigation';

@Component({
    components: {
        DashboardIcon,
        ApiKeysIcon,
        TeamIcon,
        EditProjectDropdown,
    },
})
export default class NavigationArea extends Vue {
    /**
     * Array of navigation links with icons.
     */
    public readonly navigation: NavigationLink[] = [
        RouteConfig.ProjectDashboard.withIcon(DashboardIcon),
        RouteConfig.ApiKeys.withIcon(ApiKeysIcon),
        RouteConfig.Users.withIcon(TeamIcon),
    ];

    /**
     * Indicates if navigation side bar is hidden.
     */
    public get isNavigationHidden(): boolean {
        return this.isOnboardingTour || this.isCreateProjectPage;
    }

    /**
     * Indicates if current route is create project page.
     */
    private get isCreateProjectPage(): boolean {
        return this.$route.name === RouteConfig.CreateProject.name;
    }

    /**
     * Indicates if current route is onboarding tour.
     */
    private get isOnboardingTour(): boolean {
        return this.$route.name === RouteConfig.OnboardingTour.name;
    }
}
</script>

<style scoped lang="scss">
    .navigation-svg-path {
        fill: rgb(53, 64, 73);
    }

    .navigation-area {
        padding: 25px;
        min-width: 170px;
        max-width: 170px;
        background: #e6e9ef;
        display: flex;
        flex-direction: column;
        align-items: center;
        font-family: 'font_regular', sans-serif;

        &__item-container {
            flex: 0 0 auto;
            padding: 10px;
            width: calc(100% - 20px);
            display: flex;
            justify-content: flex-start;
            align-items: center;
            margin-bottom: 40px;
            text-decoration: none;

            &__link {
                display: flex;
                justify-content: flex-start;
                align-items: center;

                &__title {
                    font-size: 16px;
                    line-height: 23px;
                    color: #1b2533;
                    margin: 0 0 0 15px;
                }
            }

            &.router-link-active,
            &:hover {
                font-family: 'font_bold', sans-serif;
                background: #0068dc;
                border-radius: 6px;

                .navigation-area__item-container__link__title {
                    color: #fff;
                }

                .svg .navigation-svg-path:not(.white) {
                    fill: #fff !important;
                    opacity: 1;
                }
            }
        }
    }
</style>
