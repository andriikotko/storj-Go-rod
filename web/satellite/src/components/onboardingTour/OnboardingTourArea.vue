// Copyright (C) 2020 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="tour-area">
        <div class="tour-area__info-bar" v-show="isInfoBarVisible && isPaywallEnabled">
            <div class="tour-area__info-bar__message">
                <b class="tour-area__info-bar__message__bold">Try Tardigrade with 50 GB Free after adding a payment method.</b>
                <p class="tour-area__info-bar__message__regular"> Cancel before your credit runs out and youâ€™ll never be billed.</p>
            </div>
            <CloseImage class="tour-area__info-bar__close-img" @click="disableInfoBar"/>
        </div>
        <div class="tour-area__content">
            <ProgressBar
                :is-paywall-enabled="isPaywallEnabled"
                :is-add-payment-step="isAddPaymentState"
                :is-create-project-step="isCreateProjectState"
                :is-create-api-key-step="isCreatApiKeyState"
                :is-upload-data-step="isUploadDataState"
            />
            <OverviewStep
                v-if="isDefaultState && isPaywallEnabled"
                @setAddPaymentState="setAddPaymentState"
            />
            <OverviewStepNoPaywall
                v-if="isDefaultState && !isPaywallEnabled"
                @setCreateProjectState="setCreateProjectState"
            />
            <AddPaymentStep
                v-if="isAddPaymentState"
                @setProjectState="setCreateProjectState"
            />
            <CreateProjectStep
                v-if="isCreateProjectState"
                @setApiKeyState="setCreateApiKeyState"
            />
            <CreateApiKeyStep
                v-if="isCreatApiKeyState"
                @setUploadDataState="setUploadDataState"
            />
            <UploadDataStep v-if="isUploadDataState"/>
            <img
                v-if="isAddPaymentState"
                class="tour-area__content__tardigrade"
                src="@/../static/images/onboardingTour/tardigrade.png"
                alt="tardigrade image"
            >
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import ProgressBar from '@/components/onboardingTour/ProgressBar.vue';
import AddPaymentStep from '@/components/onboardingTour/steps/AddPaymentStep.vue';
import CreateApiKeyStep from '@/components/onboardingTour/steps/CreateApiKeyStep.vue';
import CreateProjectStep from '@/components/onboardingTour/steps/CreateProjectStep.vue';
import OverviewStep from '@/components/onboardingTour/steps/OverviewStep.vue';
import OverviewStepNoPaywall from '@/components/onboardingTour/steps/OverviewStepNoPaywall.vue';
import UploadDataStep from '@/components/onboardingTour/steps/UploadDataStep.vue';

import CheckedImage from '@/../static/images/common/checked.svg';
import CloseImage from '@/../static/images/onboardingTour/close.svg';

import { RouteConfig } from '@/router';
import { TourState } from '@/utils/constants/onboardingTourEnums';

@Component({
    components: {
        OverviewStepNoPaywall,
        UploadDataStep,
        CreateApiKeyStep,
        CreateProjectStep,
        AddPaymentStep,
        ProgressBar,
        OverviewStep,
        CheckedImage,
        CloseImage,
    },
})

export default class OnboardingTourArea extends Vue {
    public areaState: number = TourState.DEFAULT;
    public isInfoBarVisible: boolean = true;

    /**
     * Lifecycle hook after initial render.
     * Sets area to needed state.
     */
    public mounted(): void {
        if (this.userHasProject && this.userHasApiKeys) {
            try {
                this.$router.push(RouteConfig.ProjectDashboard.path);
            } catch (error) {
                return;
            }

            return;
        }

        if (this.userHasProject && !this.userHasApiKeys) {
            this.disableInfoBar();
            this.setCreateApiKeyState();

            return;
        }

        if (this.$store.state.paymentsModule.creditCards.length > 0) {
            this.disableInfoBar();
            this.setCreateProjectState();

            return;
        }

        if (this.$store.getters.isTransactionProcessing || this.$store.getters.isBalancePositive) {
            this.setAddPaymentState();
        }
    }

    /**
     * Indicates if paywall is enabled.
     */
    public get isPaywallEnabled(): boolean {
        return this.$store.state.paymentsModule.isPaywallEnabled;
    }

    /**
     * Indicates if area is in default state.
     */
    public get isDefaultState(): boolean {
        return this.areaState === TourState.DEFAULT;
    }

    /**
     * Indicates if area is in adding payment method state.
     */
    public get isAddPaymentState(): boolean {
        return this.areaState === TourState.ADDING_PAYMENT;
    }

    /**
     * Indicates if area is in creating project state.
     */
    public get isCreateProjectState(): boolean {
        return this.areaState === TourState.PROJECT;
    }

    /**
     * Indicates if area is in api key state.
     */
    public get isCreatApiKeyState(): boolean {
        return this.areaState === TourState.API_KEY;
    }

    /**
     * Indicates if area is in upload data state.
     */
    public get isUploadDataState(): boolean {
        return this.areaState === TourState.UPLOAD;
    }

    /**
     * Sets area's state to adding payment method state.
     */
    public setAddPaymentState(): void {
        this.areaState = TourState.ADDING_PAYMENT;
    }

    /**
     * Sets area's state to creating project state.
     */
    public setCreateProjectState(): void {
        this.disableInfoBar();
        this.areaState = TourState.PROJECT;
    }

    /**
     * Sets area's state to creating api key state.
     */
    public setCreateApiKeyState(): void {
        this.areaState = TourState.API_KEY;
    }

    /**
     * Sets area's state to upload data state.
     */
    public setUploadDataState(): void {
        this.areaState = TourState.UPLOAD;
    }

    /**
     * Disables info bar visibility.
     */
    public disableInfoBar(): void {
        this.isInfoBarVisible = false;
    }

    /**
     * Indicates if user has at least one project.
     */
    private get userHasProject(): boolean {
        return this.$store.state.projectsModule.projects.length > 0;
    }

    /**
     * Indicates if user has at least one API key.
     */
    private get userHasApiKeys(): boolean {
        return this.$store.state.apiKeysModule.page.apiKeys.length > 0;
    }
}
</script>

<style scoped lang="scss">
    .tour-area {
        width: 100%;

        &__info-bar {
            display: flex;
            align-items: center;
            justify-content: space-between;
            width: calc(100% - 60px);
            padding: 10px 30px;
            background-color: #7c8794;

            &__message {
                display: flex;
                align-items: center;

                &__bold,
                &__regular {
                    margin: 0 10px 0 0;
                    font-family: 'font_regular', sans-serif;
                    font-size: 14px;
                    line-height: 21px;
                    color: #fff;
                    word-break: break-word;
                }
            }

            &__close-img {
                cursor: pointer;
                min-width: 18px;
            }
        }

        &__content {
            padding: 0 100px 80px 100px;
            position: relative;

            &__tardigrade {
                position: absolute;
                left: 50%;
                bottom: 0;
                transform: translate(-50%);
            }
        }
    }

    @media screen and (max-width: 1550px) {

        .tour-area {

            &__content {
                padding: 0 50px 80px 50px;
            }
        }
    }

    @media screen and (max-width: 1000px) {

        .tour-area {

            &__content {
                padding: 0 25px 80px 25px;
            }
        }
    }
</style>
