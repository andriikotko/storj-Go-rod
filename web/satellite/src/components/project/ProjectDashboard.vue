// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <div class="dashboard-area">
        <h1 class="dashboard-area__title">{{projectName}} Dashboard</h1>
        <ProjectUsage/>
        <ProjectSummary/>
        <BucketArea/>
    </div>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';

import BucketArea from '@/components/project/buckets/BucketArea.vue';
import ProjectSummary from '@/components/project/summary/ProjectSummary.vue';
import ProjectUsage from '@/components/project/usage/ProjectUsage.vue';

import { RouteConfig } from '@/router';
import { APP_STATE_ACTIONS } from '@/utils/constants/actionNames';
import { SegmentEvent } from '@/utils/constants/analyticsEventNames';
import { MetaUtils } from '@/utils/meta';

@Component({
    components: {
        BucketArea,
        ProjectUsage,
        ProjectSummary,
    },
})
export default class ProjectDashboard extends Vue {
    /**
     * Lifecycle hook after initial render.
     * Segment tracking is processed.
     */
    public mounted(): void {
        if (!this.$store.getters.selectedProject.id) {
            this.$router.push(RouteConfig.OnboardingTour.path);

            return;
        }

        const projectLimit: number = this.$store.getters.user.projectLimit;
        if (projectLimit && this.$store.getters.projectsCount < projectLimit) {
            this.$store.dispatch(APP_STATE_ACTIONS.SHOW_CREATE_PROJECT_BUTTON);
        }

        this.$segment.track(SegmentEvent.PROJECT_VIEWED, {
            project_id: this.$store.getters.selectedProject.id,
        });
    }

    /**
     * Returns selected project name.
     */
    public get projectName(): string {
        return this.$store.getters.selectedProject.name;
    }

    /**
     * Returns project limits increase request url from config.
     */
    public get projectLimitsIncreaseRequestURL(): string {
        return MetaUtils.getMetaContent('project-limits-increase-request-url');
    }
}
</script>

<style scoped lang="scss">
    .dashboard-area {
        padding: 50px 30px 30px 30px;
        font-family: 'font_regular', sans-serif;

        &__title {
            font-family: 'font_bold', sans-serif;
            font-size: 22px;
            line-height: 27px;
            color: #384b65;
            margin: 0 0 30px 0;
        }
    }
</style>
