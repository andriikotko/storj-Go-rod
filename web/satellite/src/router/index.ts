// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Vue from 'vue';
import Router, { RouteRecord } from 'vue-router';

import AccountArea from '@/components/account/AccountArea.vue';
import AccountBilling from '@/components/account/billing/BillingArea.vue';
import DetailedHistory from '@/components/account/billing/depositAndBillingHistory/DetailedHistory.vue';
import CreditsHistory from '@/components/account/billing/freeCredits/CreditsHistory.vue';
import SettingsArea from '@/components/account/SettingsArea.vue';
import ApiKeysArea from '@/components/apiKeys/ApiKeysArea.vue';
import Page404 from '@/components/errors/Page404.vue';
import OnboardingTourArea from '@/components/onboardingTour/OnboardingTourArea.vue';
import CreateProject from '@/components/project/CreateProject.vue';
import EditProjectDetails from '@/components/project/EditProjectDetails.vue';
import ProjectDashboard from '@/components/project/ProjectDashboard.vue';
import ProjectMembersArea from '@/components/team/ProjectMembersArea.vue';

import store from '@/store';
import { NavigationLink } from '@/types/navigation';
const DashboardArea = () => import('@/views/DashboardArea.vue');
const ForgotPassword = () => import('@/views/forgotPassword/ForgotPassword.vue');
const LoginArea = () => import('@/views/login/LoginArea.vue');
const RegisterArea = () => import('@/views/register/RegisterArea.vue');

Vue.use(Router);

/**
 * RouteConfig contains information about all routes and subroutes
 */
export abstract class RouteConfig {
    // root paths
    public static Root = new NavigationLink('/', 'Root');
    public static Login = new NavigationLink('/login', 'Login');
    public static Register = new NavigationLink('/signup', 'Register');
    public static ForgotPassword = new NavigationLink('/forgot-password', 'Forgot Password');
    public static Account = new NavigationLink('/account', 'Account');
    public static ProjectDashboard = new NavigationLink('/project-dashboard', 'Dashboard');
    public static Users = new NavigationLink('/project-members', 'Users');
    public static ApiKeys = new NavigationLink('/api-keys', 'API Keys');
    public static OnboardingTour = new NavigationLink('/onboarding-tour', 'Onboarding Tour');
    public static CreateProject = new NavigationLink('/create-project', 'Create Project');
    public static EditProjectDetails = new NavigationLink('/edit-project-details', 'Edit Project Details');

    // child paths
    public static Settings = new NavigationLink('settings', 'Settings');
    public static Billing = new NavigationLink('billing', 'Billing');
    public static BillingHistory = new NavigationLink('billing-history', 'Billing History');
    public static DepositHistory = new NavigationLink('deposit-history', 'Deposit History');
    public static CreditsHistory = new NavigationLink('credits-history', 'Credits History');
    // TODO: disabled until implementation
    // public static Referral = new NavigationLink('referral', 'Referral');

    // not in project yet
    // public static Referral = new NavigationLink('//ref/:ids', 'Referral');
}

export const notProjectRelatedRoutes = [
    RouteConfig.Login.name,
    RouteConfig.Register.name,
    RouteConfig.ForgotPassword.name,
    RouteConfig.Billing.name,
    RouteConfig.BillingHistory.name,
    RouteConfig.DepositHistory.name,
    RouteConfig.CreditsHistory.name,
    RouteConfig.Settings.name,
    // RouteConfig.Referral.name,
];

export const router = new Router({
    mode: 'history',
    routes: [
        {
            path: RouteConfig.Login.path,
            name: RouteConfig.Login.name,
            component: LoginArea,
        },
        {
            path: RouteConfig.Register.path,
            name: RouteConfig.Register.name,
            component: RegisterArea,
        },
        {
            path: RouteConfig.ForgotPassword.path,
            name: RouteConfig.ForgotPassword.name,
            component: ForgotPassword,
        },
        {
            path: RouteConfig.Root.path,
            meta: {
                requiresAuth: true,
            },
            component: DashboardArea,
            children: [
                {
                    path: RouteConfig.Account.path,
                    name: RouteConfig.Account.name,
                    component: AccountArea,
                    children: [
                        {
                            path: RouteConfig.Settings.path,
                            name: RouteConfig.Settings.name,
                            component: SettingsArea,
                        },
                        {
                            path: RouteConfig.Billing.path,
                            name: RouteConfig.Billing.name,
                            component: AccountBilling,
                        },
                        {
                            path: RouteConfig.BillingHistory.path,
                            name: RouteConfig.BillingHistory.name,
                            component: DetailedHistory,
                        },
                        {
                            path: RouteConfig.DepositHistory.path,
                            name: RouteConfig.DepositHistory.name,
                            component: DetailedHistory,
                        },
                        {
                            path: RouteConfig.CreditsHistory.path,
                            name: RouteConfig.CreditsHistory.name,
                            component: CreditsHistory,
                        },
                        // {
                        //     path: RouteConfig.Referral.path,
                        //     name: RouteConfig.Referral.name,
                        //     component: ReferralArea,
                        // },
                    ],
                },
                {
                    path: RouteConfig.ProjectDashboard.path,
                    name: RouteConfig.ProjectDashboard.name,
                    component: ProjectDashboard,
                },
                {
                    path: RouteConfig.Root.path,
                    name: 'default',
                    component: ProjectDashboard,
                },
                {
                    path: RouteConfig.Users.path,
                    name: RouteConfig.Users.name,
                    component: ProjectMembersArea,
                },
                {
                    path: RouteConfig.ApiKeys.path,
                    name: RouteConfig.ApiKeys.name,
                    component: ApiKeysArea,
                },
                {
                    path: RouteConfig.OnboardingTour.path,
                    name: RouteConfig.OnboardingTour.name,
                    component: OnboardingTourArea,
                },
                {
                    path: RouteConfig.CreateProject.path,
                    name: RouteConfig.CreateProject.name,
                    component: CreateProject,
                },
                {
                    path: RouteConfig.EditProjectDetails.path,
                    name: RouteConfig.EditProjectDetails.name,
                    component: EditProjectDetails,
                },
            ],
        },
        {
            path: '*',
            name: '404',
            component: Page404,
        },
    ],
});

router.beforeEach((to, from, next) => {
    if (navigateToDefaultSubTab(to.matched, RouteConfig.Account)) {
        next(RouteConfig.Account.with(RouteConfig.Billing).path);

        return;
    }

    if (to.name === 'default') {
        next(RouteConfig.ProjectDashboard.path);

        return;
    }

    next();
});

router.afterEach(({name}, from) => {
    if (!name) {
        return;
    }

    if (notProjectRelatedRoutes.includes(name)) {
        document.title = `${router.currentRoute.name} | ${store.state.appStateModule.satelliteName}`;

        return;
    }

    const selectedProjectName = store.state.projectsModule.selectedProject.name ?
        `${store.state.projectsModule.selectedProject.name} | ` : '';

    document.title = `${selectedProjectName + router.currentRoute.name} | ${store.state.appStateModule.satelliteName}`;
});

/**
 * if our route is a tab and has no sub tab route - we will navigate to default subtab.
 * F.E. /account/ -> /account/billing/;
 * @param routes - array of RouteRecord from vue-router
 * @param next - callback to process next route
 * @param tabRoute - tabNavigator route
 */
function navigateToDefaultSubTab(routes: RouteRecord[], tabRoute: NavigationLink): boolean {
    return routes.length === 2 && (routes[1].name as string) === tabRoute.name;
}
