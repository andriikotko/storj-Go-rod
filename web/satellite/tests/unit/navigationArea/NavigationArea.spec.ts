// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import Router from 'vue-router';
import Vuex from 'vuex';

import NavigationArea from '@/components/navigation/NavigationArea.vue';
import OnboardingTourArea from '@/components/onboardingTour/OnboardingTourArea.vue';
import ProjectDashboard from '@/components/project/ProjectDashboard.vue';

import { RouteConfig } from '@/router';
import { makeProjectsModule, PROJECTS_MUTATIONS } from '@/store/modules/projects';
import { NavigationLink } from '@/types/navigation';
import { Project } from '@/types/projects';
import { createLocalVue, shallowMount } from '@vue/test-utils';

import { ProjectsApiMock } from '../mock/api/projects';

const api = new ProjectsApiMock();
api.setMockProjects([new Project('1')]);
const projectsModule = makeProjectsModule(api);
const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(Router);

const store = new Vuex.Store({ modules: { projectsModule } });

const expectedLinks: NavigationLink[] = [
    RouteConfig.ProjectDashboard,
    RouteConfig.ApiKeys,
    RouteConfig.Users,
];

describe('NavigationArea', () => {
    it('snapshot not changed during onboarding tour', (): void => {
        const router = new Router({
            mode: 'history',
            routes: [{
                path: '/',
                name: RouteConfig.OnboardingTour.name,
                component: OnboardingTourArea,
            }],
        });
        const wrapper = shallowMount(NavigationArea, {
            store,
            localVue,
            router,
        });

        const navigationElements = wrapper.findAll('.navigation-area__item-container');

        expect(navigationElements.length).toBe(0);
        expect(wrapper).toMatchSnapshot();
    });

    const router = new Router({
        mode: 'history',
        routes: [{
            path: '/',
            name: RouteConfig.ProjectDashboard.name,
            component: ProjectDashboard,
        }],
    });

    it('snapshot not changed with project', async () => {
        const projects = await store.dispatch('fetchProjects');
        store.commit(PROJECTS_MUTATIONS.SELECT_PROJECT, projects[0].id);

        const wrapper = shallowMount(NavigationArea, {
            store,
            localVue,
            router,
        });

        const navigationElements = wrapper.findAll('.navigation-area__item-container');

        expect(navigationElements.length).toBe(3);
        expect(wrapper).toMatchSnapshot();
    });

    it('navigation links are correct', () => {
        const wrapper = shallowMount(NavigationArea, {
            store,
            localVue,
            router,
        });

        const navigationLinks = (wrapper.vm as any).navigation;

        expect(navigationLinks.length).toBe(expectedLinks.length);

        expectedLinks.forEach((link, i) => {
            expect(navigationLinks[i].name).toBe(expectedLinks[i].name);
            expect(navigationLinks[i].path).toBe(expectedLinks[i].path);
        });
    });
});
