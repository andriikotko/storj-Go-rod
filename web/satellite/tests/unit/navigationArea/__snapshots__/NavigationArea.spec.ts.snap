// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationArea snapshot not changed during onboarding tour 1`] = ``;

exports[`NavigationArea snapshot not changed with project 1`] = `
<div class="navigation-area">
  <editprojectdropdown-stub></editprojectdropdown-stub>
  <router-link-stub to="/project-dashboard" tag="a" ariacurrentvalue="page" event="click" aria-label="Dashboard" class="navigation-area__item-container">
    <div class="navigation-area__item-container__link">
      <anonymous-stub></anonymous-stub>
      <h1 class="navigation-area__item-container__link__title">Dashboard</h1>
    </div>
  </router-link-stub>
  <router-link-stub to="/api-keys" tag="a" ariacurrentvalue="page" event="click" aria-label="API Keys" class="navigation-area__item-container">
    <div class="navigation-area__item-container__link">
      <anonymous-stub></anonymous-stub>
      <h1 class="navigation-area__item-container__link__title">API Keys</h1>
    </div>
  </router-link-stub>
  <router-link-stub to="/project-members" tag="a" ariacurrentvalue="page" event="click" aria-label="Users" class="navigation-area__item-container">
    <div class="navigation-area__item-container__link">
      <anonymous-stub></anonymous-stub>
      <h1 class="navigation-area__item-container__link__title">Users</h1>
    </div>
  </router-link-stub>
</div>
`;
